### PR TITLE
Bulk Insert: Reuse indexes for unique_by

### PR DESCRIPTION
**Pull request by kaspth**
I found `:unique_by` with `:columns` and `:where` inside it tough to
grasp. The documentation only mentioned indexes and partial indexes.
So why duplicate a model's indexes in an insert_all/upsert_all call
when we can just look it up?

This has the added benefit of raising if no index is found, such that
people can't insert thousands of records without relying on an index
of some form.

cc @boblail

### CODE REVIEW

File: activerecord/lib/active_record/insert_all.rb [hunk 1]
```diff
@@ -69,14 +76,17 @@ def readonly_columns
...
 
       def unique_by_columns
-        unique_by ? unique_by.fetch(:columns).map(&:to_s) : []
+        Array(unique_by_index&.columns)
       end
 
-      def primary_keys
-        Array.wrap(model.primary_key)
+      def find_index_for(unique_by_index)
+        if index = connection.indexes(model.table_name).index_by(&:name)[unique_by_index.to_s]
```
Comment (kaspth) [inline activerecord/lib/active_record/insert_all.rb]: @kamipo is this okay to do? Or will it execute a query for the schema?

Comment (kaspth) [reply activerecord/lib/active_record/insert_all.rb]: Another thing we could do is matching on the columns ala: `find { |index| index.columns.sort == unique_by_index }` to get `unique_by_index: :isbn` and `unique_by_index: %i( author_id name )`.

**Comment by kaspth** [comment]
Another benefit is that if an index is changed, users won't have to hunt down `insert_all`/`upsert_all` calls in their app 😊

File: activerecord/lib/active_record/insert_all.rb [hunk 2]
```diff
@@ -116,9 +126,10 @@ def returning
         end
 
         def conflict_target
-          return unless conflict_columns
-          sql = +"(#{quote_columns(conflict_columns).join(',')})"
-          sql << " WHERE #{where}" if where
+          index = insert_all.unique_by_index
+
+          sql = +"(#{quote_columns(index&.columns || insert_all.primary_keys).join(',')})"
```
Comment (vishaltelangre) [inline activerecord/lib/active_record/insert_all.rb]: I think we need to set primary keys as conflict target only if `update_duplicates?` is true.

File: activerecord/lib/active_record/insert_all.rb (hunk 1 above)
Comment (kamipo) [reply activerecord/lib/active_record/insert_all.rb]: I think it is a good timing to make the index cache on `schema_cache`.
https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/schema_cache.rb

File: activerecord/lib/active_record/insert_all.rb [hunk 3]
```diff
@@ -29,6 +31,11 @@ def updatable_columns
       keys - readonly_columns - unique_by_columns
     end
 
+    def primary_keys
+      Array(model.primary_key)
+    end
+
+
```
Comment (kamipo) [inline activerecord/lib/active_record/insert_all.rb]: Are this double blank lines intentional?

File: activerecord/lib/active_record/insert_all.rb (hunk 2 above)
Comment (boblail) [reply activerecord/lib/active_record/insert_all.rb]: Good catch, @vishaltelangre. Yeah, that's true. It's OK (desirable, even) to leave the **conflict target** clause off of `ON CONFLICT DO NOTHING`. Implicitly using primary keys here would cause duplicates to be skipped (by default) _only_ if they were duplicate by primary key.

Comment (boblail) [inline activerecord/lib/active_record/insert_all.rb]: We can also provide the name of an index as **conflict target**! so:

> Example:
> ```sql
> INSERT ... ON CONFLICT index_books_on_isbn DO UPDATE ...
> ```

If we _only_ support identifying indexes by name, I'm pretty sure this method boils down to

```ruby
def conflict_target
  insert_all.unique_by_index
end
```

**Review by boblail — COMMENTED** [review]
:+1: Cool! I think supporting identifying the **conflict target** by name is a solid improvement!

(Also, I didn't know about `connection.indexes(model.table_name)` — TIL!)

I've got two questions about the API:

### unique_by_index

is kind of a wordy parameter name. Do we need to add `_index`? On one hand, it helps to signal to users what value is expected. On the other hand, Rails-generated index names start with `"index_"` so the phrases `unique_by_index: "index_books_on_isbn"` repeats the word "index", and the phrase `unique_by: "index_books_on_isbn"` reads just fine.

### Support both?

Does it make sense to support _both_ ways of specifying an index? We could reuse [remove_index](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-remove_index)'s API here to increase predictability:

> Examples:
> ```ruby
> Book.insert_all books, unique_by: { name: "index_books_on_isbn" }
>```
> ```ruby
> Book.insert_all books, unique_by: { column: :isbn }
>```
> ```ruby
> Book.insert_all books, unique_by: { column: %i[author title] }
> ```


If wanted to support both, but departed a little from `remove_index`'s API, we could forego the nested hash by treating a string/symbol value as an index name and an array value as an array of columns:

> Examples:
> ```ruby
> Book.insert_all books, unique_by: "index_books_on_isbn"
>```
> ```ruby
> Book.insert_all books, unique_by: %i[:isbn]
>```
> ```ruby
> Book.insert_all books, unique_by: %i[author title]
>```

(Note: in this scenario, we're dropping support for `:where` — and removing one way of specifying a partial index — but we've added the ability to specify a partial index by name. The array-of-columns syntax just gives you a less-wordy way of describing the most commonplace indexes.)

Lastly, should we loop @dhh in on the conversation? He weighed in pretty heavily on the API of the original PR.

**Comment by boblail** [comment]
BTW, thanks for looping me in again, @kaspth! 😄

File: activerecord/lib/active_record/insert_all.rb (hunk 2 above)
Comment (kaspth) [reply activerecord/lib/active_record/insert_all.rb]: Yes, I had a version of this that looked like the below but I cut `update_duplicates?` here as connection's `build_insert_sql` already guarded with `update_duplicates?`. However, I see now that would still fuck it up for `skip_duplicates?`, right. I think this would be clearer if it was split into `conflict_target_for_skip` and `conflict_target_for_update`. I'll experiment, thanks.

```ruby
def conflict_target
  if index = insert_all.unique_by_index
    +"(#{quote_columns(index.columns).join(',')})".tap do |sql|
      sql << " WHERE #{index.where}" if index.where
    end
  elsif update_duplicates?
    "(#{quote_columns(insert_all.primary_keys).join(',')})"
  end
end
```

File: activerecord/lib/active_record/insert_all.rb (hunk 3 above)
Comment (kaspth) [reply activerecord/lib/active_record/insert_all.rb]: Yes! 😊

File: activerecord/lib/active_record/insert_all.rb (hunk 1 above)
Comment (kaspth) [reply activerecord/lib/active_record/insert_all.rb]: Great, I gotta try that!

File: activerecord/lib/active_record/insert_all.rb (hunk 2 above)
Comment (kaspth) [reply activerecord/lib/active_record/insert_all.rb]: Screenshot 2019-03-09 at 20 48 51

Comment (kaspth) [reply activerecord/lib/active_record/insert_all.rb]: Hm, I cannot seem to get that to work. And I can't find documentation for either SQLIte3, Postgres or MySQL that say they support it. How does it work?

**Comment by kaspth** [comment]
True, the parameter name is too wordy. I've been thinking we just go with `unique_index: :isbn` with the column shorthand. That alludes to the fact that `insert_all` defaults to check duplicates on the primary key and you're merely changing that default.

Though it is tough to say which expect should be stressed more. Is it the duplicate check or is it the aspect that the check is done with an index? I do like the latter because it hints of the performance aspect e.g. if you're inserting 100.000 rows into an already large table, you'd want to check the index performance first.

File: activerecord/lib/active_record/insert_all.rb [hunk 4]
```diff
@@ -69,13 +75,23 @@ def readonly_columns
...
-        unique_by ? unique_by.fetch(:columns).map(&:to_s) : []
+        Array(unique_index&.columns)
       end
 
-      def primary_keys
-        Array.wrap(model.primary_key)
+      def find_index_for(unique_index)
+        match = Array(unique_index).map(&:to_s).sort
+
+        if index = indexes.find { |i| match.include?(i.name) || i.columns.sort == match }
```
Comment (kamipo) [inline activerecord/lib/active_record/insert_all.rb]: Now we could verify whether or not a unique index.

```suggestion
        if index = indexes.find { |i| i.unique && (match.include?(i.name) || i.columns.sort == match) }
```

Comment (kaspth) [reply activerecord/lib/active_record/insert_all.rb]: Great point! I thought about and forgot to do it 😄

File: activerecord/lib/active_record/insert_all.rb [hunk 5]
```diff
@@ -69,13 +75,23 @@ def readonly_columns
       end
 
       def unique_by_columns
-        unique_by ? unique_by.fetch(:columns).map(&:to_s) : []
+        Array(unique_index&.columns)
       end
 
-      def primary_keys
-        Array.wrap(model.primary_key)
+      def find_index_for(unique_index)
```
Comment (kaspth) [inline activerecord/lib/active_record/insert_all.rb]: Should probably be `find_unique_index_for`.

File: activerecord/lib/active_record/insert_all.rb [hunk 6]
```diff
@@ -69,13 +75,23 @@ def readonly_columns
       end
 
       def unique_by_columns
```
Comment (kaspth) [inline activerecord/lib/active_record/insert_all.rb]: Should be `unique_index_columns`.

File: activerecord/lib/active_record/insert_all.rb (hunk 2 above)
Comment (boblail) [reply activerecord/lib/active_record/insert_all.rb]: 🤔 I might've read https://www.postgresql.org/docs/9.5/sql-insert.html#SQL-ON-CONFLICT wrong. But does it work like this? (with `ON CONSTRAINT`)

```sql
INSERT ... ON CONFLICT ON CONSTRAINT index_books_on_isbn DO UPDATE ...
```

Then again, maybe that's not the best plan:

> Tip: It is often preferable to use unique index inference rather than naming a constraint directly using `ON CONFLICT ON CONSTRAINT constraint_name`. Inference will continue to work correctly when the underlying index is replaced by another more or less equivalent index in an overlapping way, for example when using `CREATE UNIQUE INDEX ... CONCURRENTLY` before dropping the index being replaced.

Comment (kaspth) [reply activerecord/lib/active_record/insert_all.rb]: I tried that but it threw an "Unknown constraint error".

Ah nice tidbit! So that's why they prefer to infer the index instead of naming it.

**Comment by kaspth** [comment]
I'm pretty happy with this. I'm still open to other names and I'm going to do a final documentation pass tomorrow.

I've also killed the sort I had on the `columns` when finding indexes. That order means something so I don't think an `[ :author_id, :name ]` index should we found with `[ :name, :author_id ]`.

**Review by boblail — APPROVED** [review]

**Comment by kaspth** [comment]
Went back to our first love: `unique_by`! Also did a documentation pass while I was digging through the options documentation.

@boblail I was honestly very surprised that insert_all/upsert_all matches every unique index by default. I thought it was just the primary key index that was used! And that `unique_by` would then simply shift to using a different index. Also makes it seem somewhat strange that passing a unique_by can make insert_all/upsert_all throw an `ActiveRecord::RecordNotUnique` error. You'd think that would be for `insert_all!`. So we should probably also look into the raising aspects of these two variants.

**Comment by simi** [comment]
:+1: for keeping `unique_by`
